### DISCUSSION TITLE
Companion App Mode Guardian CGM to XDrip. no data coming in

### DISCUSSION THREAD

**Discussion by illuzionist** [NightscoutFoundation/xDrip#3025] [Q&A]
hi. I'm trying to configure Xdrip+ for local data transfer from Guardian CGM. but the data is not transferred to Xdrip+.

I did the following: installed the data source "Companion Application" in Xdrip+
gave permission to view Xdrip + alerts.
I left only two constant alerts with sugar data - Xdrip+ and Guardian CGM
stopped the sensor and started the sensor again. according to this instruction. https://xdrip.readthedocs.io/en/latest/install/companion/#allow-notification-access-greyed-out
but the data does not arrive in Xdrip+. I have attached a screenshot of the error notification. What am I doing wrong?
I have already downloaded the new version and installed it on the phone, I have not imported the old settings, I have configured everything again, but it does not work. samsung a71 phone, android 13
![Screenshot_20230815_200402_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/4f2f44fd-12f0-40c7-bf47-3f050311fbc8)
![Screenshot_20230814_161250_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/571d9488-e4d8-4bfa-a25f-a0922ea14a32)

**Comment by Navid200** [comment]
Can you post a screenshot of a notification from your Guardian app?

**Reply by illuzionist** [reply]
if you translate the inscription from the Guardian cgm notification into English: "glucose sensor".
or do I need to take a screenshot from another place in the app?thank you!
![Screenshot_20230815_204302_One UI Home](https://github.com/NightscoutFoundation/xDrip/assets/138001377/9f07675e-4f5b-4e11-9d7d-7f7aafd7fa8e)

**Reply by Navid200** [reply]
That's what I wanted to see.
Can we do an experiment?  This is only an experiment, not necessarily a solution.

Can you change the settings of your Guardian app so that it would use English language so that the glucose readings would be shown like the following?
6.9 mmol/l

**Reply by illuzionist** [reply]
there is no item about the interface language in the application itself. changed the language in the phone settings.
![Screenshot_20230815_210800_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/830644f1-6bae-4406-9f48-2315c7911a5a)
![Screenshot_20230815_210804_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/98d9be1d-2599-489d-a520-be797f6911b8)

**Reply by Navid200** [reply]
Is xDrip showing readings now?

**Reply by illuzionist** [reply]
no(. after changing the language, I had to issue permission to determine geolocation.
![Screenshot_20230815_211808_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/397e1135-be21-4fd8-a69d-f4e610d1724d)
![Screenshot_20230815_211816_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/50ae7c06-c925-4d8e-80e0-f36adf118e7a)

**Reply by Navid200** [reply]
According to your logs, xDrip is to set to CarelInk follower, not Companion app.

**Reply by illuzionist** [reply]
yes, that's right. it was on time until 20.59 according to the logs. at 21.03 there was an error that appears in my Companion app mode.
![Screenshot_20230815_212712_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/3ef8354c-65f1-41db-bdc3-6ac071d57d18)
![Screenshot_20230815_212725_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/7b1c7d70-14e8-4d0d-90d1-381edf0c4a65)

**Reply by Navid200** [reply]
Does xDrip have notification access shown on the list as shown here?
https://navid200.github.io/xDrip/docs/Follow/CompanionApp.html

**Reply by illuzionist** [reply]
yes
![Screenshot_20230815_213704_Settings](https://github.com/NightscoutFoundation/xDrip/assets/138001377/f47b15d1-324f-44c9-991f-9db70c2474e6)

**Reply by Navid200** [reply]
Thanks for reporting this.
I have asked for advice.

**Reply by ah7675** [reply]
@illuzionist Are you using Guardian 4 standalone? I suspect the package name is not in xDrip+ whitelist.

https://github.com/NightscoutFoundation/xDrip/discussions/3005#discussioncomment-6751348

**Reply by illuzionist** [reply]
@ah7675 I use Enlite sensors + guardian connect transmitter + Guardian CGM app

**Reply by ah7675** [reply]
Sorry may not able to help. I'll leave it to other experts.
But at least from log you shared in https://github.com/NightscoutFoundation/xDrip/discussions/3005#discussioncomment-6751348, I suspect that xDrip+ either not receive notification or can't recognize package name.

**Reply by PureblackCD** [reply]
Hi, I have the same issue. Xdrip cannot extract data from Guardian Connect notifications. 
Phone: Samsung A34, Android 13
Sensor: Medtronic G3
Application: Guardian Connect 3.4.3 + Xdrip+ - [Nightly build 10th Aug 2023]

The same printscreen with xdrip notifications access enabled and companion app enabled. 


Thank you, 
Cristi

**Comment by jamorham** [comment]
Can anyone provide a link to the play store for exactly which app is being used here?

**Reply by PureblackCD** [reply]
xdrip+ is not in playstore. the link is: https://github.com/NightscoutFoundation/xDrip/releases
guardian connect: https://play.google.com/store/apps/details?id=com.medtronic.diabetes.guardianconnect&hl=en_US&pli=1

thank you, 
Cristi

**Reply by ah7675** [reply]
Actually, I tried debug by myself yesterday.
However, Guardian connect shows notification only when you were paired with the transmitter. 

@jamorham I'm wondering how do you usually debug issue like this? Whenever you don't have specific CGM.

**Reply by benceszasz** [reply]
@jamorham 

**Medtronic Guardian Connect standalone CGM android apps (used with Enlite and Guardian 3 sensors)**
US: com.medtronic.diabetes.guardianconnect.us
Outside US: com.medtronic.diabetes.guardianconnect

**Medtronic Guardian standalone CGM android apps (used with Guardian 4 sensors)**
US: com.medtronic.diabetes.guardianconnect.us
Outside US: com.medtronic.diabetes.guardian

**Guardian Connect glucose notification text** 
The notification text of the Guardian Connect app is a little tricky, it contains two non readable characters:
NO-BREAK SPACE (u00a0)
WORD JOINER (u2060)
The NO-BREAK SPACE (u00a0) is placed after the SG value and the WORD JOINER (u2060) is after the "/".
For example the displayed "8,9 mmol\l" value is actually: \u0038\u002c\u0039\u00a0\u006d\u006d\u006f\u006c\u2060\u002f\u2060\u006c\u0020
I think that the current filterString method of UiBasedCollector may not be able to handle it.
These special characters are not used in the Guardian app's notification, only in Guardian Connect.

**Reply by jamorham** [reply]
@benceszasz this is extremely detailed and helpful thank you

**Comment by Navid200** [comment]
@illuzionist Please respond to JamOrHam's request.

**Comment by jamorham** [comment]
@illuzionist and any other interested users, please try the latest nightly to see if this resolves the issue?

**Reply by illuzionist** [reply]
@jamorham @Navid200 @benceszasz and other. 
it works! thank you very much!
![Screenshot_20230819_000953_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/5531dcfa-9923-4cea-9e94-81e015a4f03c)
![Screenshot_20230819_001005_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/fa0182ba-ac2e-4dde-9d4c-b7e867a85b5e)
![Screenshot_20230819_001035_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/8649cd2f-a37a-4052-acbc-df03e5cf53ee)

**Reply by illuzionist** [reply]
I'm back after a few days of testing the Nightly build version on 26th Aug 2023. 
I noticed two interruptions within 3-4 days. 
the first case was on absolutely identical sugar indicators, which lasted for about 30 minutes. one or two readings were missed, but then, when the sugar readings changed up or down, Xdrip+ started downloading data. There are no screenshots.
the second case - Please see the screenshots. 
Thank you.

![Screenshot_20230830_104014_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/9de956b5-86d4-444c-bfad-0e841a42c675)
![Screenshot_20230830_103952_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/cc363f7f-570c-4ed5-98e9-cde20a6e95a7)
![Screenshot_20230830_104056_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/f6ccae6d-cf55-422c-adca-b4823f1c8556)

**Comment by PureblackCD** [comment]
Also working for me.Thank you!
![Screenshot_20230819_002933_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/91929759/f384846c-248e-4fec-b53d-07e07da9b00c)

**Comment by PureblackCD** [comment]
I have to complete: after a few readings, xdrip+ stopped receiving data and resumed to receive after 4-5 readings (25 mins).
This happened to someone else? Or just me?

Thank you!
Cristi 


![Screenshot_20230819_013335_xDrip+](https://github.com/NightscoutFoundation/xDrip/asset
![Screenshot_20230819_013353_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/91929759/e
![Screenshot_20230819_013329_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/91929759/7bab9ff9-909a-48d8-aaa7-54797745d341)
9a61364-1b2e-4c82-aa7b-2c5fc0cc2361)
s/91929759/02b4e174-145e-4de3-a732-66b4575b0674)

**Reply by benceszasz** [reply]
Do you still experience missing values, when UiBasedCollector reports "Did not find any matches" in the log?
Turn on debug log for UiBasedCollector in Menu> Settings > Less Common Settings > Extra Logging Settings > Extra tags for logging add UiBasedCollector:d
By turning on debug log a lot ore information will be available in the logs.
Share the UiBasedCollector debug log of the unsuccessful Notification data extractions ("Did not find any matches").

**Reply by illuzionist** [reply]
the data stops coming in. after that, the entry "MissedReadingService" appears in the log file (21.57 is the time of the error in the screenshot). after a while, the data starts coming in again.

![Screenshot_20230819_220515_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/ba65a5d0-2311-48a2-afa7-1af6bc60d833)
![Screenshot_20230819_220443_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/5b4294ff-dbdc-4938-a313-e8a83e5defd2)
![Screenshot_20230819_220504_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/89b4bf34-c74e-4ecb-9efd-24d20a401597)

**Reply by illuzionist** [reply]
I removed the previous version and installed the latest "Nightly build 19th Aug 2023". if I understand correctly, then there is an entry in the log file about the last sugar reading (the same value in Guardian CGM), but it is not on the main screen in Xdrip+.

![Screenshot_20230820_005552_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/54a1cfa8-cef3-43d2-b9d5-29a62f78d4a5)
![Screenshot_20230820_005513_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/da84dbc9-3ec0-4e08-81a8-22c9be2a3918)
![Screenshot_20230820_005533_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/a95d5b6d-92d9-4741-a1a2-ed5a9b1cfd36)

**Reply by Mishu12111** [reply]
Hi! I also have the same problem. I tried to add in  Menu> Settings > Less Common Settings > Extra Logging Settings > Extra tags for logging to add UiBasedCollector:d . It works but there still are interruptions in readings for about 15-20 minutes between.  xDrip+ version: Nightly build 19th Aug 2023
![1](https://github.com/NightscoutFoundation/xDrip/assets/89389790/28c2bd92-c9e1-4bb9-bbde-186b1d103833)

**Reply by jamorham** [reply]
Please try the latest nightly to see if it resolves the issue of missed readings.

**Reply by illuzionist** [reply]
downloaded the latest version yesterday. there were no errors yesterday, but this morning the data disappeared for 15 minutes. changed the data source in the settings (conpanionApp-CareLinc Follower-conpanionApp) then the data was loaded via conpanionApp.

![Screenshot_20230824_094536_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/f86833f6-44b9-4ec5-bc70-15fa65494917)
![Screenshot_20230824_094559_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/89014502-f213-4e9d-b52b-172620165771)
![Screenshot_20230824_094610_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/1e341ead-5f03-4896-86d7-4036a0d35632)

**Reply by illuzionist** [reply]
now I have managed to catch this error once again

![Screenshot_20230826_023008_Guardian CGM](https://github.com/NightscoutFoundation/xDrip/assets/138001377/54164c48-b014-4124-a01a-305bf6e3dbf2)
![Screenshot_20230826_022953_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/d2a8dc4e-cfc1-4d46-bea8-7e50aa86c39b)
![Screenshot_20230826_022938_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/138001377/21ab9844-8df7-4395-bd0d-ab029888f6a9)

**Reply by ah7675** [reply]
@jamorham From my understanding, there is a restriction and assume sensor values won't be the same for most 3 or 4 times (jammed). I think that's a kind of workaround for specific CGM app. 
At least for G7 & Guardian 3/4, the digi number will be removed from notification upon signal lost or sensor error.
Is it possible to make "jammed" rule applies to specific app only?

**Reply by jamorham** [reply]
@ah7675 Its a safety precaution in case of either of the two unlikely events of the co-opted app misbehaving/failing or us interpreting a static value which is not glucose. For apps which apply considerable smoothing its possible that this could be triggered incorrectly as you show above. I will extend it slightly by default and put a match for guardian to make the period more like an hour.  I'll be interested to know how that performs.

**Reply by jamorham** [reply]
@illuzionist @ah7675 There is a new nightly which changes the jam detection period. Please let me know if this helps.

**Reply by Mishu12111** [reply]
Hi! I am using Medtronic Guardian for sensor 4 and the latest version for xDrip+(Nightly build 26th Aug 2023) and do not get BG readings using Companion App. xDrip+ has the same level of access as Aaps which includes notification access and so on. I am not complaining, I am just providing information from my own use experience!
![WhatsApp Image 2023-08-12 at 21 54 53](https://github.com/NightscoutFoundation/xDrip/assets/89389790/68de09ff-c1a0-4db8-8e52-e3d251f926ee)
![WhatsApp Image 2023-08-12 at 21 57 03](https://github.com/NightscoutFoundation/xDrip/assets/89389790/87f28894-295d-43ad-9eb5-5fc6201eed72)
![WhatsApp Image 2023-08-26 at 14 27 47 (1)](https://github.com/NightscoutFoundation/xDrip/assets/89389790/cb56c4a0-0f7a-4a1f-9fdc-e8d58a796c1f)
![WhatsApp Image 2023-08-26 at 14 27 47](https://github.com/NightscoutFoundation/xDrip/assets/89389790/85b4efb2-b5ed-44e6-a66a-d532e89a70bc)
![WhatsApp Image 2023-08-26 at 14 27 48](https://github.com/NightscoutFoundation/xDrip/assets/89389790/d295a759-5439-4dea-8627-75699b04991b)

**Reply by benceszasz** [reply]
Did you turn on debug log for UiBasedCollector in Menu> Settings > Less Common Settings > Extra Logging Settings > Extra tags for logging and by adding UiBasedCollector:d
By turning on debug log more information will be available in the logs for UIBasedCollector

**Reply by Mishu12111** [reply]
Yes, I have already added UiBasedCollector:d but it still does not work.
![WhatsApp Image 2023-08-27 at 10 26 14 (1)](https://github.com/NightscoutFoundation/xDrip/assets/89389790/9ae67a35-f1fc-4da7-abe7-20bdd713613f)
![WhatsApp Image 2023-08-27 at 10 26 14](https://github.com/NightscoutFoundation/xDrip/assets/89389790/dd69cfa6-9553-4805-a241-b07c6013b789)

**Reply by jamorham** [reply]
@Mishu12111 I think I know what might cause this I will prepare a test version for you to try.

**Reply by Mishu12111** [reply]
> @Mishu12111 I think I know what might cause this I will prepare a test version for you to try.

Ok,thank you very much!

**Reply by Mishu12111** [reply]
@jamorham 
I have installed the newest update (Nightly build 27th Aug 2023). I will test it today and give feedback!
Thank you!
From recent testing, I can see that it is able to read the BG using Companion App, but it is behind with one reading: in xDrip+ I have as the last reading 95 while in Medtronic the last reading is 97. 
![WhatsApp Image 2023-08-27 at 15 16 36](https://github.com/NightscoutFoundation/xDrip/assets/89389790/d7a7b921-9d90-45c4-ad2a-6227e6b119a4)

![WhatsApp Image 2023-08-27 at 15 15 45 (1)](https://github.com/NightscoutFoundation/xDrip/assets/89389790/0060057d-8ae9-487a-901f-a7c385c3a18f)
![WhatsApp Image 2023-08-27 at 15 15 45](https://github.com/NightscoutFoundation/xDrip/assets/89389790/91acf4d8-73fa-42df-aa1e-5f3d30c28898)

**Reply by jamorham** [reply]
@Mishu12111 please try the latest nightly and let me know if that helps

**Reply by jamorham** [reply]
@Mishu12111 so I only just saw your previous message. So did this new nightly fix the problem you were having before? How frequently does the guardian sensor provide new readings (to understand the problem you reported of xDrip being behind?)

**Reply by Mishu12111** [reply]
Yes,this update solved the previous problem. Now i receive BG readings from Medtronic for sensor 4 notification. The new problem is that it falls behind with one reading for some reason. xDrip+ is a reading behind.
Medtronic for sensor 4 gives a new reading for every 5 minutes.

**Reply by Mishu12111** [reply]
@jamorham 
After some testing of the newest xDrip+ version(Nightly build 27th Aug 2023): The previous problem is solved! Now I get BG readings from Medtronic for sensor 4 using Companion App! Thank you very much! 
The new problem is that, as I said, xDrip+ has a delay of a reading. For instance, this morning I had the value of 108 on Medtronic while on xDrip+ I had 114, which was the previous value Medtronic provided. I can see that in the Events Log is read the current BG but on the main screen is shown the previous one.
![WhatsApp Image 2023-08-28 at 11 48 10 (1)](https://github.com/NightscoutFoundation/xDrip/assets/89389790/5e821583-f803-402d-b967-49741fbceeda)
![WhatsApp Image 2023-08-28 at 11 48 10 (2)](https://github.com/NightscoutFoundation/xDrip/assets/89389790/4db0a0af-4ce3-4d29-b77e-9d75ebb39d28)
![WhatsApp Image 2023-08-28 at 11 48 10](https://github.com/NightscoutFoundation/xDrip/assets/89389790/c9051ade-1ee4-46fc-9298-2a13df0ee139)
![WhatsApp Image 2023-08-28 at 11 48 11 (1)](https://github.com/NightscoutFoundation/xDrip/assets/89389790/aae87d48-78d6-49fe-a911-fae9bb7c8855)
![WhatsApp Image 2023-08-28 at 11 48 11](https://github.com/NightscoutFoundation/xDrip/assets/89389790/5488699d-539c-44f4-8ac0-2c88221d9fe3)

**Reply by ah7675** [reply]
Since notification doesn't provide the timestamp of reading, xDrip+ can only use the first reading time as sync-point then get next reading for every 4m30s.
Actually, Medtronic/Dexcom app refresh notification about every 1 minute, hence edge trigger update won't work neither.

**Reply by jamorham** [reply]
This can be fixed but the logic is a little complicated so it might take a little while before it is implemented but I will see what I can do.

**Reply by jamorham** [reply]
@Mishu12111 actually one work-around you could try might go like this: Turn phone off for 20 minutes. Turn phone back on, see if when the readings come in again they are then in sync.

**Reply by Mishu12111** [reply]
@jamorham I have tried it(to turn the phone off for 20 minutes) but it still doesn't work, the readings are still not synced.

**Comment by jamorham** [comment]
Is it that the missing readings occur when the official app is reporting that it cannot reach the sensor but there is no back filling capability for this data source within xDrip?

**Comment by ah7675** [comment]
I believe it's due to trend arrow characters after glucose value/unit.

**Reply by benceszasz** [reply]
@ah7675 yes, the trend arrows are also not filtered from the notification text thus the conversion of the notification text to glucose value fails.

@jamorham since there can be many unwanted characters in the glucose notification text, wouldn't it be easier to revert the filtering logic in the `filterString` method of the `UiBasedCollector` and try to keep only the relevant characters/texts? 
For example
```
// keep only characters, numbers, decimal separators, relevant signs
.replaceAll("[^\\w, \\-\\.]", "")
// make it case insensitive
.toUpperCase()
// remove glucose units
.replace("MGDL", " ")
.replace("MMOLL", " ")
.trim();
```

By keeping the letter characters could make it sure, that if there is an active insulin notification text with insulin unit in case of a pump's companion app, then it wouldn't be translated to a glucose value, for example "5.4 U" wouldn't be translated to "5.4" glucose (in mmol/l).

What do you think?

**Reply by jamorham** [reply]
@benceszasz we need to be very careful that we don't create something which fails and accepts invalid data in the future. I would prefer to adopt the strategy that we interpret all the notifications that we know about and we refuse to interpret anything that is unknown. The collector is designed to be flexible so if we need in future to adopt a broader approach for any specific app we can easily do that. I will produce an update to hopefully deal safely with the trend arrows and we can go from there.

**Comment by Mishu12111** [comment]
Feedback issue! With Medtronic Guardian for senzor 3 i had readings but with many interruptions (which lasted 15-20 minutes) but when I tried with Medtronic 4 I didn't had readings at all. I use the special app: Medtronic Guardian for sensor 4. I am not unthankful or complaining, I am just providing information from my own experience of use! 
![WhatsApp Image 2023-08-22 at 12 03 22 (1)](https://github.com/NightscoutFoundation/xDrip/assets/89389790/eb5670a8-ffde-4dbb-9690-f9892f660125)
![WhatsApp Image 2023-08-22 at 12 03 22](https://github.com/NightscoutFoundation/xDrip/assets/89389790/f750af08-f0e1-4517-8450-0896afd96364)

**Comment by prometey87** [comment]
Carelink Follower
Hasn't been working for a few days now.  EU server.
![Screenshot_20230824_065509_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/44647470/0a587400-8b81-4935-8b92-453755527e2e)
![Uploading 
![Screenshot_20230824_065457_xDrip+](https://github.com/NightscoutFoundation/xDrip/assets/44647470/79a32859-6e0f-4387-a4c9-4716ea423fdf)
Screenshot_20230824_065500_xDrip+.jpg…]()

**Reply by prometey87** [reply]
Pump 780G, sensor G4.

**Reply by PureblackCD** [reply]
You have to use Companion app not carelink follower..

**Comment by PureblackCD** [comment]
Good morning!
I installed yesterday the last version, compiled yesterday and everything is smooth . No interruptions  AT ALL! (17 hours from installation)  Everything is in order. 
Xdrip is set to read CGM notifications with Companion app and to upload data to Nightscout. The carelink servers are shunted so their malfunction does not affect me anymore. 
Thank you, @jamorham !

**Reply by Mishu12111** [reply]
Hi! What kind of sensor are you using? 3 or 4?

**Reply by PureblackCD** [reply]
Hi, the sensor is G3

**Reply by Mishu12111** [reply]
Ok,thank you! With sensor 4 there are no readings with Companion App.

**Reply by Troxelke** [reply]
I am using companion app with Guardian 4 sensors (interface Minimed Mobile app) and works just great!

So happy Carelink isn't in the loop anymore!
